### DISCUSSION TITLE
How to access collections in a WebC component?

### DISCUSSION THREAD

**Discussion by helenclx** [11ty/eleventy#3248] [Q&A]
Hi, I am trying to refactor [my portfolio website that was built in Eleventy](https://github.com/helenclx/helenclx.github.io). However, I cannot figure out how to access my collection in a WebC component.

I want to refactor my [header](https://github.com/helenclx/helenclx.github.io/blob/main/src/_includes/main/header-main.webc), which is a WebC component to render links by looping through [my "pages" collection](https://github.com/helenclx/helenclx.github.io/tree/main/src/pages). I read the [`webc:for` Loops](https://www.11ty.dev/docs/languages/webc/#webcfor-loops) section of the WebC article on Eleventy's official documentation, so I am trying to do something like this in the HTML portion of my header WebC component:

When I tried to search in the Eleventy Discord server, I discovered a thread in September 2023 where someone also asked about using collections in WebC, but the only reply was a suggestion of replacing the [1, 2, 3] in the samples in the documentation with a `collections` variable, such as`collections.pages`, which did not work at all, since when I tried to build my pages, the terminal would output error messages stating the variable as undefined.
```
[11ty] Original error stack trace: Error: Check the dynamic attribute: `webc:for="collections.pages"`.
[11ty] Original error message: Cannot read properties of undefined (reading 'pages')
```
When I tried to `console.log` `collections.pages`, my browser console outputs shows `Uncaught ReferenceError: collections is not defined`.

The OP of the September 2023 Discord forum thread is no longer in the server, and the thread received no other replies, and I have also tried to search the internet, but there is a surprising lack of information and resources about accessing collections in WebC, so I am unable to find the solution.

Is there an actual way to access our collections in a WebC component?

**Comment by helenclx** [comment]
Update: I decided to make my header template to a Nunjucks file so I can iterate my pages collection to render navigation links. That said, I would still like to see more support and documentation for accessing collections in WebC.

**Comment by pdehaan** [comment]
I think the 11ty/11ty-website is written in .webc and might have some examples: https://github.com/search?q=repo%3A11ty%2F11ty-website+path%3A*.webc&type=code

**Comment by rothsandro** [comment]
In components you need to use the `$data` prefix to access global variables: `$data.collections.pages` should work.

See the [release notes](https://github.com/11ty/webc/releases/tag/v0.11.0)

**Reply by helenclx** [reply]
I tried and still had no success, unfortunately.

**Reply by caseywatts** [reply]
I came across this thread doing something similar (but different), and this comment was the hint I needed to get data from a global data file working in a webc component, thank you!

The need to use `$data` isn't documented on the webc documentation yet; I'll open an issue.
https://www.11ty.dev/docs/languages/webc/

Thank you both 🙏

**Reply by caseywatts** [reply]
I found an existing ticket about documenting `$data` and I added to that one some:
https://github.com/11ty/11ty-website/issues/1774#issue-2891744896

**Comment by helenclx** [comment]
The only WebC components I could find that access the collections were from solution-loisir's [webc-starter-kit](https://github.com/solution-loisir/webc-starter-kit), which creates collections by using [a separate config file in JavaScript](https://github.com/solution-loisir/webc-starter-kit/blob/master/config/collections.js).

**Comment by cypressSap** [comment]
A bit late for the original question, but hopefully this helps anyone looking for answers in the future:

@rothsandro is correct that you need to use the (still undocumented) `$data` prefix to access data from inside a WebC component. But it looks like the other half of the problem was from [using the component via the `renderTemplate` shortcode](https://github.com/helenclx/helenchong.dev/blob/acec1f147bf3a6272e231facd93789657b01b976/src/_includes/main/base.njk#L39) - by default, [only `eleventy` and `page` data are available inside the shortcode](https://www.11ty.dev/docs/plugins/render/#pass-in-data), and anything else has to be passed in:

**`base.njk`**
```njk
{% renderTemplate "webc", { collections: collections } %}
    <!--- Component now has `$data.collections` defined --->
    <header-main></header-main>
{% endrenderTemplate %}
```

Unfortunately this is only documented on the page for the Render plugin, not the WebC page.
